# Shrooms: Edible or Poisonous?

'Shrooming' -- the act of foraging for edible mushrooms in the forest -- is growing in popularity. However, this novel hobby presents new dangers: there are countless types of unique mushrooms, many of which can easily be misclassified. One small mistake in mushroom indentification can have enormous consequences. Accidentally ingesting a poisonous mushroom can lead to serious medical complications, and perhaps even death. In this notebook, I'll attempt to use data science methodologies to correctly identify a mushroom as either edible or poisonous.

In [1]:
# Start by making key imports: NumPy and Pandas
import numpy as np
import pandas as pd

In [2]:
# Read in CSV file of our data
shrooms = pd.read_csv('../datasets/mushrooms.csv')

In [3]:
# Use the .describe() attribute for initial analysis
shrooms.describe()

class cap-shape cap-surface cap-color bruises  odor gill-attachment  \
count   8124      8124        8124      8124    8124  8124            8124   
unique     2         6           4        10       2     9               2   
top        e         x           y         n       f     n               f   
freq    4208      3656        3244      2284    4748  3528            7914   

       gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
count          8124      8124       8124  ...                     8124   
unique            2         2         12  ...                        4   
top               c         b          b  ...                        s   
freq           6812      5612       1728  ...                     4936   

       stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1          4   
top                         w                      w         p          w   
freq                     4464                   4384      8124       7924   

       ring-number ring-type spore-print-color population habitat  
count         8124      8124              8124       8124    8124  
unique           3         5                 9          6       7  
top              o         p                 w          v       d  
freq          7488      3968              2388       4040    3148  

[4 rows x 23 columns]

In [4]:
# Check the number of rows and columns in the dataset
shrooms.shape

(8124, 23)

In [5]:
# Count the number of nulls in the dataset
shrooms.isnull().sum()

# No nulls!

class                       0
cap-shape                   0
cap-surface                 0
cap-color                   0
bruises                     0
odor                        0
gill-attachment             0
gill-spacing                0
gill-size                   0
gill-color                  0
stalk-shape                 0
stalk-root                  0
stalk-surface-above-ring    0
stalk-surface-below-ring    0
stalk-color-above-ring      0
stalk-color-below-ring      0
veil-type                   0
veil-color                  0
ring-number                 0
ring-type                   0
spore-print-color           0
population                  0
habitat                     0
dtype: int64

In [6]:
# Check the data types of the dataset. All the values in the datasets are objects
shrooms.dtypes

class                       object
cap-shape                   object
cap-surface                 object
cap-color                   object
bruises                     object
odor                        object
gill-attachment             object
gill-spacing                object
gill-size                   object
gill-color                  object
stalk-shape                 object
stalk-root                  object
stalk-surface-above-ring    object
stalk-surface-below-ring    object
stalk-color-above-ring      object
stalk-color-below-ring      object
veil-type                   object
veil-color                  object
ring-number                 object
ring-type                   object
spore-print-color           object
population                  object
habitat                     object
dtype: object

In [7]:
# For each column, display the name of the column and the number of unique attributes for each column
for column in shrooms:
    print(column)
    print(shrooms[column].unique())

class
['p' 'e']
cap-shape
['x' 'b' 's' 'f' 'k' 'c']
cap-surface
['s' 'y' 'f' 'g']
cap-color
['n' 'y' 'w' 'g' 'e' 'p' 'b' 'u' 'c' 'r']
bruises
['t' 'f']
odor
['p' 'a' 'l' 'n' 'f' 'c' 'y' 's' 'm']
gill-attachment
['f' 'a']
gill-spacing
['c' 'w']
gill-size
['n' 'b']
gill-color
['k' 'n' 'g' 'p' 'w' 'h' 'u' 'e' 'b' 'r' 'y' 'o']
stalk-shape
['e' 't']
stalk-root
['e' 'c' 'b' 'r' '?']
stalk-surface-above-ring
['s' 'f' 'k' 'y']
stalk-surface-below-ring
['s' 'f' 'y' 'k']
stalk-color-above-ring
['w' 'g' 'p' 'n' 'b' 'e' 'o' 'c' 'y']
stalk-color-below-ring
['w' 'p' 'g' 'b' 'n' 'e' 'y' 'o' 'c']
veil-type
['p']
veil-color
['w' 'n' 'o' 'y']
ring-number
['o' 't' 'n']
ring-type
['p' 'e' 'l' 'f' 'n']
spore-print-color
['k' 'n' 'u' 'h' 'w' 'r' 'o' 'y' 'b']
population
['s' 'n' 'a' 'v' 'y' 'c']
habitat
['u' 'g' 'm' 'd' 'p' 'w' 'l']


#### Need for dummy variables

Looking at the unique values in each column of the dataset, we can see that there are a lot of unique non-numerical values in the shrooms dataset. This means we're going to have to create a new column for each unique variable for each unique column. These will be `dummy variables` that can be used for quantitative analysis.

How many dummy variables, exactly? Let's take a look and see

In [8]:
# Create a new dataframe that's simply a description of the shrooms dataset. 
foo = shrooms.describe()
foo

class cap-shape cap-surface cap-color bruises  odor gill-attachment  \
count   8124      8124        8124      8124    8124  8124            8124   
unique     2         6           4        10       2     9               2   
top        e         x           y         n       f     n               f   
freq    4208      3656        3244      2284    4748  3528            7914   

       gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
count          8124      8124       8124  ...                     8124   
unique            2         2         12  ...                        4   
top               c         b          b  ...                        s   
freq           6812      5612       1728  ...                     4936   

       stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1          4   
top                         w                      w         p          w   
freq                     4464                   4384      8124       7924   

       ring-number ring-type spore-print-color population habitat  
count         8124      8124              8124       8124    8124  
unique           3         5                 9          6       7  
top              o         p                 w          v       d  
freq          7488      3968              2388       4040    3148  

[4 rows x 23 columns]

In [9]:
# Now we can create a column called 'sum' and sum all the unqiue values in the dataset
foo['sum'] = foo.sum(axis = 1)
foo['sum']

count                      186852
unique                        119
top       exynfnfcbbtbsswwpwopwvd
freq                       108158
Name: sum, dtype: object

In [10]:
foo['sum']['unique'] * shrooms.shape[1]

2737

#### 119 Dummy Variables

Looking at the number above, we can see that we're going to have to create 119 dummy variables for this dataset. We'll create those variables when we come to the `Feature Engineering` portion of the notebook.

**What exactly would be considered a good model?** This question is unique to each data science problem. Let's look at the percentage of edible mushrooms in our dataset:

In [11]:
# The following number represents the percentage of edible mushrooms in our dataset:
shrooms[shrooms['class'] == 'e'].shape[0] / shrooms.shape[0]

0.517971442639094

#### 51.79% of mushrooms in our dataset at edible.

0.5179 indicates the number of edible mushrooms in our dataset. This might lead some to say that any model with an accuracy score above 51.79% is superior, as it outperforms the natural percentage of edible and poisonous mushrooms in the dataset. However, the consequences of eating a poisonous mushroom are dire: even one misclassification can result in death. This is a good reminder that every data science problem is unique and should be approached as such.

## Feature Engineering

The biggest (and perhaps only) amount of feature engineering required in this problem is the creation of the `dummy variables` for the mushroom dataset.

The easiest way to create all the `dummy variables` is to create a function that can create all the variables and thgen drop the original column from the `shrooms` dataset. Once the dummies have been created, the original dataset is no longer needed.

In [12]:
# Function to dummify the shrooms dataframe
def dummify(df):
    
    # Create new, empty dataframe, which we'll append dummy variables to
    df2 = pd.DataFrame()
    
    # Create a for loop to dummify shrooms columns 1 by 1
    for column in df:
        
        # Create the dummies unique to one column
        dummy = pd.get_dummies(df[column], prefix = column, drop_first = True)
        
        # Add dummies to df2 column by column
        for i in dummy:
            df2[i] = dummy[i]
            
    # Return the new dataframe        
    return df2

In [13]:
shrooms = dummify(shrooms)
shrooms

class_p  cap-shape_c  cap-shape_f  cap-shape_k  cap-shape_s  \
0           1            0            0            0            0   
1           0            0            0            0            0   
2           0            0            0            0            0   
3           1            0            0            0            0   
4           0            0            0            0            0   
...       ...          ...          ...          ...          ...   
8119        0            0            0            1            0   
8120        0            0            0            0            0   
8121        0            0            1            0            0   
8122        1            0            0            1            0   
8123        0            0            0            0            0   

      cap-shape_x  cap-surface_g  cap-surface_s  cap-surface_y  cap-color_c  \
0               1              0              1              0            0   
1               1              0              1              0            0   
2               0              0              1              0            0   
3               1              0              0              1            0   
4               1              0              1              0            0   
...           ...            ...            ...            ...          ...   
8119            0              0              1              0            0   
8120            1              0              1              0            0   
8121            0              0              1              0            0   
8122            0              0              0              1            0   
8123            1              0              1              0            0   

      ...  population_n  population_s  population_v  population_y  habitat_g  \
0     ...             0             1             0             0          0   
1     ...             1             0             0             0          1   
2     ...             1             0             0             0          0   
3     ...             0             1             0             0          0   
4     ...             0             0             0             0          1   
...   ...           ...           ...           ...           ...        ...   
8119  ...             0             0             0             0          0   
8120  ...             0             0             1             0          0   
8121  ...             0             0             0             0          0   
8122  ...             0             0             1             0          0   
8123  ...             0             0             0             0          0   

      habitat_l  habitat_m  habitat_p  habitat_u  habitat_w  
0             0          0          0          1          0  
1             0          0          0          0          0  
2             0          1          0          0          0  
3             0          0          0          1          0  
4             0          0          0          0          0  
...         ...        ...        ...        ...        ...  
8119          1          0          0          0          0  
8120          1          0          0          0          0  
8121          1          0          0          0          0  
8122          1          0          0          0          0  
8123          1          0          0          0          0  

[8124 rows x 96 columns]

## Exploratory Data Analysis

Now that the data have been quantified, we can do more exploratory data analysis on these quantities before creating a predictive model.

In [14]:
# Our new dataframe has 8,124 rows and 96 columns
shrooms.shape

(8124, 96)

In [15]:
# A list of the columns in the dataframe 'shrooms'
shrooms.columns

Index(['class_p', 'cap-shape_c', 'cap-shape_f', 'cap-shape_k', 'cap-shape_s',
       'cap-shape_x', 'cap-surface_g', 'cap-surface_s', 'cap-surface_y',
       'cap-color_c', 'cap-color_e', 'cap-color_g', 'cap-color_n',
       'cap-color_p', 'cap-color_r', 'cap-color_u', 'cap-color_w',
       'cap-color_y', 'bruises_t', 'odor_c', 'odor_f', 'odor_l', 'odor_m',
       'odor_n', 'odor_p', 'odor_s', 'odor_y', 'gill-attachment_f',
       'gill-spacing_w', 'gill-size_n', 'gill-color_e', 'gill-color_g',
       'gill-color_h', 'gill-color_k', 'gill-color_n', 'gill-color_o',
       'gill-color_p', 'gill-color_r', 'gill-color_u', 'gill-color_w',
       'gill-color_y', 'stalk-shape_t', 'stalk-root_b', 'stalk-root_c',
       'stalk-root_e', 'stalk-root_r', 'stalk-surface-above-ring_k',
       'stalk-surface-above-ring_s', 'stalk-surface-above-ring_y',
       'stalk-surface-below-ring_k', 'stalk-surface-below-ring_s',
       'stalk-surface-below-ring_y', 'stalk-color-above-ring_c',
       'stalk-colo

In [28]:
def count_averages(df):
    df2 = pd.DataFrame()
    df2['variable'] = [column for column in df]
    df2['average'] = [df[column].mean() for column in df]
    return df2

averages = count_averages(shrooms)
averages

variable   average
0       class_p  0.482029
1   cap-shape_c  0.000492
2   cap-shape_f  0.387986
3   cap-shape_k  0.101920
4   cap-shape_s  0.003939
..          ...       ...
91    habitat_l  0.102413
92    habitat_m  0.035943
93    habitat_p  0.140817
94    habitat_u  0.045298
95    habitat_w  0.023634

[96 rows x 2 columns]

In [34]:
averages.sort_values(by = 'average', ascending = False).head(10)

variable   average
69                veil-color_w  0.975382
27           gill-attachment_f  0.974151
71               ring-number_o  0.921713
47  stalk-surface-above-ring_s  0.637125
50  stalk-surface-below-ring_s  0.607582
41               stalk-shape_t  0.567208
58    stalk-color-above-ring_w  0.549483
66    stalk-color-below-ring_w  0.539636
88                population_v  0.497292
76                 ring-type_p  0.488429

These letters don't do us much good to better understand our data. How are we supposed to know what each letter means?
Luckily, the dataset came with a table that explains each value in more detail. In the cell below, I've created a dictionary of each term as it corresponds to each letter, which we can then use to change the column names in our dataset

In [ ]:
# Create a dictionary

new_column_names = {
    'class_p' : 'poisonous',
    'cap-shape_c' : 'cap_shape_conical',
    'cap-shape_f' : 'cap_shape_flat',
    'cap-shape_k' : 'cap_shape_knobbed',
    'cap-shape_s' : 'cap_shape_sunken',
    'cap-shape_x' : 'cap_shape_convex',
    'cap-surface_g' : 'cap_surface_grooves',
    'cap-surface_s' : 'cap_surface_smooth',
    'cap-surface_y' : 'cap_surface_scaly',
    'cap-color_c' : 'cap_color_cinnamon',
    'cap-color_e' : 'cap_color_red',
    'cap-color_g' : 'cap_color_grey',
    'cap-color_n' : 'cap_color_brown',
    'cap-color_p' : 'cap_color_pink',
    'cap-color_r' : 'cap_color_green',
    'cap-color_u' : 'cap_color_purple',
    'cap-color_w' : 'cap_color_white',
    'cap-color_y' : 'cap_color_yellow',
    'bruises_t' : 'bruises_true',
    'odor_c' : 'odor_creosote',
    'odor_f' : 'odor_foul',
    'odor_l' : 'odor_anise',
    'odor_m' : 'odor_musty',
    'odor_n' : 'odor_none',
    'odor_p' : 'odor_pungent',
    'odor_s' : 'odor_spicy',
    'odor_y' : 'odor_fishy',
    'gill-attachment_f' : 'gill_attachment_free',
    'gill-spacing_w' : 'gill_attachment_crowded'
    
    
    
}

**Just by looking at the above dataframe, we can see that 97.5% of the mushrooms in our dataset have a "Veil Color" that's categorized as "W", 97.4% have a Gill Attachment that's categorized as "F", and 92% have a "Ring Number" that's categorized as "O"**

Let's import some data visualization libraries for some further EDA

In [35]:
import seaborn as sns
import matplotlib.pyplot as plt